File: go/result.go
package sdk

import (
	"context"
	"encoding/json"
	grpc "github.com/crawlab-team/crawlab-grpc"
	"github.com/crawlab-team/crawlab-sdk/entity"
	"github.com/crawlab-team/crawlab-sdk/interfaces"
	"github.com/crawlab-team/go-trace"
)

var RS = NewResultService()

type ResultService struct {
	// internals
	sub grpc.TaskService_SubscribeClient
}

func (svc *ResultService) SaveItem(items ...entity.Result) {
	svc.save(items)
}

func (svc *ResultService) SaveItems(items []entity.Result) {
	svc.save(items)
}

func (svc *ResultService) save(items []entity.Result) {
	// skip if no task id specified
	if GetTaskId().IsZero() {
		return
	}

	var records []interface{}
	for _, item := range items {
		records = append(records, item)
	}
	data, err := json.Marshal(&entity.StreamMessageTaskData{
		TaskId:  GetTaskId(),
		Records: records,
	})
	if err != nil {
		trace.PrintError(err)
		return
	}
	if err := svc.sub.Send(&grpc.StreamMessage{
		Code: grpc.StreamMessageCode_INSERT_DATA,
		Data: data,
	}); err != nil {
		trace.PrintError(err)
		return
	}
}

func (svc *ResultService) init() (err error) {
	svc.sub, err = C.GetTaskClient().Subscribe(context.Background())
	if err != nil {
		return trace.TraceError(err)
	}
	return nil
}

func NewResultService(opts ...ResultServiceOption) interfaces.ResultService {
	// service
	svc := &ResultService{}

	// apply options
	for _, opt := range opts {
		opt(svc)
	}

	// initialize
	if err := svc.init(); err != nil {
		panic(err)
	}

	return svc
}

func SaveItem(items ...entity.Result) {
	RS.SaveItem(items...)
}

func SaveItems(items []entity.Result) {
	RS.SaveItems(items)
}

File: python/crawlab/result.py
import json
from typing import List, Optional

from crawlab.client import get_client, Client
from crawlab.config import get_task_id

from crawlab.entity.result import Result
from crawlab.grpc.entity.stream_message_code_pb2 import INSERT_DATA
from crawlab.grpc.entity.stream_message_pb2 import StreamMessage
from crawlab.grpc.services.task_service_pb2_grpc import TaskServiceStub


class ResultService:
    # internal
    c: Client = None
    task_stub: TaskServiceStub = None

    def __init__(self):
        self.c = get_client()
        self.task_stub = self.c.task_service_stub

    def save_item(self, *items: Result):
        self.save(list(items))

    def save_items(self, items: List[Result]):
        self.save(items)

    def save(self, items: List[Result]):
        _items: List[Result] = []
        for i, item in enumerate(items):
            _items.append(item)
            if i > 0 and i % 50 == 0:
                self._save(_items)
        if len(_items) > 0:
            self._save(_items)

    def _save(self, items: List[Result]):
        # task id
        tid = get_task_id()
        if tid is None:
            return

        records = []
        for item in items:
            item.set_task_id(tid)
            records.append(item)

        data = json.dumps({
            "task_id": tid,
            "records": records,
        }).encode('utf-8')

        msg = self.yield_msg(StreamMessage(
            code=INSERT_DATA,
            data=data,
        ))
        try:
            self.task_stub.Subscribe(msg)
        except Exception as e:
            print(e)
        finally:
            pass

    @staticmethod
    def yield_msg(msg):
        yield msg


RS: Optional[ResultService] = None


def get_result_service() -> ResultService:
    global RS
    if RS is not None:
        return RS
    RS = ResultService()
    return RS


def save_item(*items: Result):
    get_result_service().save_item(*items)


def save_items(items: List[Result]):
    get_result_service().save_items(items)
